README.md
express-better-ratelimit
========================

Express.js request rate limit middleware by IP with MemoryStore


history.md
## v1.1.2 / November 27, 2014
- bump to `jscs >= 1.8.0` and `jscs-doc >= 0.2.0`
- update semver range

## v1.1.1 / November 27, 2014
- fix code style collisions

## v1.1.0 / November 27, 2014
- enforce new code style
  - add jshint, jshintignore, jscs (using jscs-jsdoc)
  - normalizing all @tunnckoCore packages
  - add/edit dotfiles 
- option `message_429` deprecated, it will be removed in **v1.2.x**, use `accessLimited`
- option `message_403` deprecated, it will be removed in **v1.2.x**, use `accessForbidden`
- _>=v1.2.x jscs: "requireCamelCaseOrUpperCaseIdentifiers: true"_
- _>=v1.2.x jshint: "camelcase: true"_

## v1.0.2 / November 14, 2014
- add more dotfiles, lol
- clean unnessecery `if`

## v1.0.1 / October 27, 2014
- update readme, github

## v1.0.0 / October 27, 2014
- initial release


[npmjs-url]: http://npm.im/express-better-ratelimit
[npmjs-shields]: http://img.shields.io/npm/v/express-better-ratelimit.svg
[npmjs-install]: https://nodei.co/npm/express-better-ratelimit.svg?mini=true

[coveralls-url]: https://coveralls.io/r/tunnckoCore/express-better-ratelimit?branch=master
[coveralls-shields]: https://img.shields.io/coveralls/tunnckoCore/express-better-ratelimit.svg

[license-url]: https://github.com/tunnckoCore/express-better-ratelimit/blob/master/license.md
[license-img]: http://img.shields.io/badge/license-MIT-blue.svg

[travis-url]: https://travis-ci.org/tunnckoCore/express-better-ratelimit
[travis-img]: https://travis-ci.org/tunnckoCore/express-better-ratelimit.svg?branch=master

[depstat-url]: https://david-dm.org/tunnckoCore/express-better-ratelimit
[depstat-img]: https://david-dm.org/tunnckoCore/express-better-ratelimit.svg

[author-gittip-img]: http://img.shields.io/gittip/tunnckoCore.svg
[author-gittip]: https://www.gittip.com/tunnckoCore
[author-github]: https://github.com/tunnckoCore
[author-twitter]: https://twitter.com/tunnckoCore

[author-website]: http://www.whistle-bg.tk
[author-npmjs]: https://npmjs.org/~tunnckocore

[express-url]: https://github.com/strongloop/express

test.js
/**
 * express-better-ratelimit <https://github.com/tunnckoCore/express-better-ratelimit>
 *
 * Copyright (c) 2014 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

/**
 * Module dependencies.
 */

var express = require('express');
var limit = require('./index');
var request = require('supertest');
var PORT = 3333;

function hello(req, res, next) {
  res.set('Content-Type', 'text/html');
  res.status(200).send('<p>Hello test.</p>');
  next();
}

/**
 * app with non-default options
 */
var appNonDefault = express()
appNonDefault
  .use(limit({
    duration: 30000, //30 seconds
    max: 3, //max requests
    env: 'test'
  }))
  .use(hello)
  .listen(PORT++);

describe('app with non-default options', function() {
  it('should status 200 - 1.2.3.4 - remaining 2/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '1.2.3.4')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '2')
    .end(done);
  });
  it('should status 200 - 1.2.3.4 - remaining 1/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '1.2.3.4')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '1')
    .end(done);
  });
  it('should status 200 - 1.2.3.4 - remaining 0/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '1.2.3.4')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 429 - 1.2.3.4 - remaining 0/3 /1', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '1.2.3.4')
    .expect(429, '429: Too Many Requests.')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 200 - 8.8.8.8 - remaining 2/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '8.8.8.8')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '2')
    .end(done);
  });
  it('should status 200 - 8.8.8.8 - remaining 1/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '8.8.8.8')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '1')
    .end(done);
  });
  it('should status 429 - 1.2.3.4 - remaining 0/3 /2', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '1.2.3.4')
    .expect(429, '429: Too Many Requests.')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 200 - 8.8.8.8 - remaining 0/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '8.8.8.8')
    .expect(200, '<p>Hello test.</p>')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
  it('should status 429 - 8.8.8.8 - remaining 0/3', function(done) {
    request(appNonDefault)
    .get('/')
    .set('x-expressip', '8.8.8.8')
    .expect(429, '429: Too Many Requests.')
    .expect('X-RateLimit-Limit', '3')
    .expect('X-RateLimit-Remaining', '0')
    .end(done);
  });
});

/**
 * app with blacklisted ips
 */
var appBlack = express()
appBlack
  .use(limit({
    blackList: ['4.4.1.*'],
    message_403: 'access forbidden, please contact foo@bar.com',
    env: 'test'
  }))
  .use(hello)
  .listen(PORT++);

describe('app with blacklisted ips', function() {
  it('should 403 Forbidden - blackList', function(done) {
    request(appBlack)
    .get('/')
    .set('x-expressip', '4.4.1.8')
    .expect(403, 'access forbidden, please contact foo@bar.com')
    .end(done);
  });
});

/**
 * app with whitelisted ips
 */
var appWhite = express()
appWhite
  .use(limit({
    whiteList: ['127.0.*.*'],
    max: 50,
    env: 'test'
  }))
  .use(hello)
  .listen(PORT++);

describe('app with whitelisted ips', function() {
  it('should 200 OK - whiteList - no limits', function(done) {
    request(appWhite)
    .get('/')
    .set('x-expressip', '127.0.4.4')
    .expect(200, '<p>Hello test.</p>')
    .end(done);
  });
});


index.js
/**
 * express-better-ratelimit <https://github.com/tunnckoCore/express-better-ratelimit>
 *
 * Copyright (c) 2014 Charlike Mike Reagent, contributors.
 * Released under the MIT license.
 */

'use strict';

/**
 * Module dependencies.
 */

var ipchecker = require('ipchecker');

var defaults = {
  duration: 1000 * 60 * 60 * 24,
  whiteList: [],
  blackList: [],
  accessLimited: '429: Too Many Requests.',
  accessForbidden: '403: This is forbidden area for you.',
  max: 500,
  env: null
};

/**
 * With options through init you can control
 * black/white lists, limit per ip and reset interval.
 *
 * @param {Object} options
 * @api public
 */
module.exports = function expressBetterRatelimit(options) {
  options = options || {};

  var db = {};

  for (var key in defaults) {
    if (!options[key]) {
      options[key] = defaults[key];
    }
  }

  if (options.message_429) {
    options.accessLimited = options.message_429;
  }

  if (options.message_403) {
    options.accessForbidden = options.message_403;
  }

  var whiteListMap = ipchecker.map(options.whiteList);
  var blackListMap = ipchecker.map(options.blackList);

  return function ratelimit(req, res, next) {
    var ip = options.env === 'test' ? req.headers['x-expressip'] : req.ip;

    if (ipchecker.check(ip, blackListMap)) {
      res.status(403).send(options.accessForbidden);
      return;
    }
    if (ipchecker.check(ip, whiteListMap)) {
      return next()
    }

    var now = Date.now()
    var reset = now + options.duration;

    if (!db.hasOwnProperty(ip)) {
      db[ip] = {ip: ip, reset: reset, limit: options.max}
    }

    var delta = db[ip].reset - now;
    var retryAfter = delta / 1000 | 0;

    db[ip].limit = db[ip].limit - 1
    res.set('X-RateLimit-Limit', options.max);

    if (db[ip].reset > now) {
      res.set('X-RateLimit-Remaining', db[ip].limit < 0 ? 0 : db[ip].limit);
    }

    /* istanbul ignore if */
    if (db[ip].limit < 0 && db[ip].reset < now) {
      db[ip] = {ip: ip, reset: reset, limit: options.max};
      db[ip].limit = db[ip].limit - 1;
      res.set('X-RateLimit-Remaining', db[ip].limit);
    }
    res.set('X-RateLimit-Reset', db[ip].reset);

    if (db[ip].limit < 0) {
      res.set('Retry-After', retryAfter);
      res.status(429).send(options.accessLimited);
      return;
    }

    return next();
  };
}
